File: LeetCode/Python/solutions/p2405.py
class Solution:
    def partitionString(self, s: str) -> int:
        #
        # Solution v1: Brute Force: Pointer
        #
        # Runtime: 154 ms @ (beats) 34.43%
        # Memory Usage: 14.6 MB @ (beats) 47.10%
        #
        # ans = 1
        # chars = [-1] * 26
        # last = 0
        # for i in range(len(s)):
        #     p = ord(s[i]) - ord('a')
        #     prev = chars[p]
        #
        #     if prev >= last:
        #         ans += 1
        #         last = i
        #
        #     chars[p] = i
        #
        # return ans

        #
        # Solution v2: Brute Force: String
        #
        # Runtime: 86 ms @ (beats) 96.4%
        # Memory Usage: 14.6 MB @ (beats) 47.1%
        #
        ans = 0
        curr = s[0]
        for c in s:
            if c not in curr:
                curr += c
            else:
                ans += 1
                curr = c
        return ans


if __name__ == "__main__":
    solution = Solution()

    s = "abacaba"
    print(f"return: {solution.partitionString(s)}")  # 4

    s = "ssssss"
    print(f"return: {solution.partitionString(s)}")  # 6

    s = "cuieokbs"
    print(f"return: {solution.partitionString(s)}")  # 1

File: LeetCode/Java/src/solutions/p2405_OptimalPartitionString/P2405.java
package solutions.p2405_OptimalPartitionString;

import java.util.Arrays;

public class P2405 {
    public int partitionString(String s) {
        /*
         * Solution v1: Brute Force
         *
         * Runtime: 8 ms @ (beats) 95.45%
         * Memory Usage: 42.6 MB @ (beats) 96.17%
         */
        int ans = 1;

        int[] last_seen = new int[26];
        Arrays.fill(last_seen, -1);
        int last = 0;

        for (int i = 0; i < s.length(); i++) {
            int temp = s.charAt(i) - 'a';
            int prev = last_seen[temp];

            if (prev >= last) {
                ans++;
                last = i;
            }
            last_seen[temp] = i;

        }
        return ans;

    }
}
